File: CPP/445.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        // https://en.cppreference.com/w/cpp/container/stack
                // Stack is useful in this case : LIFO
        stack<int> numL1, numL2;
        
        while (l1){
            numL1.push(l1->val);
            l1 = l1->next;
        }
        while (l2){
            numL2.push(l2->val);
            l2 = l2->next;
        }
        //Create return ListNode
        ListNode *retNode = new ListNode(0);
        retNode->next = nullptr;
        int bit_carry = 0, sum = 0, num1 = 0, num2 = 0;
        
        while (!numL1.empty() | !numL2.empty() | bit_carry){
            if (!numL1.empty()){
                num1 = numL1.top();
                numL1.pop();
            }
            if (!numL2.empty()){
                num2 = numL2.top();
                numL2.pop();
            }
            sum = num1 + num2 + bit_carry;
            bit_carry = sum / 10;
            sum %= 10;
            num1 = 0, num2 = 0; // Reset the values -> In case each ListNode length different.
            
            ListNode *tempNode = new ListNode(sum);
            // Anchor to retNode so that retNode->tempNode2->tempNode1...
            tempNode->next = retNode->next;
            retNode->next = tempNode;
        }
        return retNode->next;        
    }
};

File: Python/445.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def addTwoNumbers(self, l1: ListNode, l2: ListNode) -> ListNode:
        stackOne = [] # Stack implementation using list
        stackTwo = []
        while l1:
            stackOne.append(l1.val)
            l1 = l1.next
        while l2:
            stackTwo.append(l2.val)
            l2 = l2.next
        # Create return ListNode
        ret = ListNode(0, None)
        carry = 0; sum = 0; num1 =0; num2 = 0
        while stackOne or stackTwo or carry:
            if stackOne:
                num1 = stackOne[-1]
                stackOne.pop()
            if stackTwo:
                num2 = stackTwo[-1]
                stackTwo.pop()

            sum = num1 + num2 + carry
            carry = sum // 10 # To avoid having floats.
            sum %= 10
            num1 = 0; num2=0
            temp = ListNode(sum, ret.next)
            ret.next = temp
        
        return ret.next
        
        
        